File: Programmers/[68646] 풍선 터트리기/Johoseong/68646.py
def solution(a):
    answer = 0
    left, right = 10 ** 10, 10 ** 10

    matrix = [[0] * len(a) for _ in range(2)]

    for i in range(len(a)):
        if left > a[i]:
            left = a[i]
        matrix[0][i] = left

    for i in range(len(a) - 1, -1, -1):
        if right > a[i]:
            right = a[i]
        matrix[1][i] = right

    for i in range(len(a)):
        if a[i] <= matrix[0][i] or a[i] <= matrix[1][i]:
            answer += 1

    return answer

File: Programmers/[68646] 풍선 터트리기/woong-jae/68646.js
function solution(a) {
  const segmentTree = Array(4 * a.length);
  
  const init = (start, end, node) => {
      if(start === end) return segmentTree[node] = start;
      
      const mid = Math.floor((start + end) / 2);
      
      const left = init(start, mid, node * 2);
      const right = init(mid + 1, end, node * 2 + 1);
      return segmentTree[node] = (a[left] < a[right] ? left : right);
  }
  
  function query(left, right) {
      function helper(start, end, node, left, right) {
          if (end < left || right < start) return null;
          if (left <= start && end <= right) return segmentTree[node];
          
          const mid = Math.floor((start + end) / 2);
          
          const leftMinIndex = helper(start, mid, node * 2, left, right);
          const rightMinIndex = helper(mid + 1, end, node * 2 + 1, left, right);
          
          if(!(leftMinIndex && rightMinIndex)) return leftMinIndex || rightMinIndex;
          
          return a[leftMinIndex] < a[rightMinIndex] ? leftMinIndex : rightMinIndex;
      }
      return helper(0, a.length - 1, 1, left, right);
  }
  
  init(0, a.length - 1, 1);
  
  const mid = query(0, a.length - 1);
  let leftEnd = mid, rightStart = mid;
 
  let result = 1;
  while(leftEnd > 0) {
      leftEnd = query(0, leftEnd - 1);
      result++;
  }
  while(rightStart < a.length - 1) {
      rightStart = query(rightStart + 1, a.length - 1);
      result++;
  }
  
  return result;
}